AirCasting/ABConnector/SDCardAirBeamServices.swift

```diff
@@ -0,0 +1,113 @@
+// Created by Lunar on 17/11/2021.
+//
+
+import CoreBluetooth
+import CoreMIDI
+
+enum SDCardData {
+    case chunk(SDCardDataChunk)
+    case metadata(SDCardMetaData)
+}
+
+struct SDCardDataChunk {
+    let payload: String
+    let sessionType: SDCardSessionType
+}
+
+struct SDCardMetaData {
+    let sessionType: SDCardSessionType
+    let measurementsCount: Int
+}
+
+protocol SDCardAirBeamServices {
+    func downloadData(from peripheral: CBPeripheral, progress: @escaping (SDCardData) -> Void, completion: @escaping (Result<Void, Error>) -> Void)
+}
+
+class BluetoothSDCardAirBeamServices: SDCardAirBeamServices {
+    private let singleChunkMeasurementsCount = 4
+    // has notifications about measurements count in particular csv file on SD card
+    private let DOWNLOAD_META_DATA_FROM_SD_CARD_CHARACTERISTIC_UUID = CBUUID(string:"0000ffde-0000-1000-8000-00805f9b34fb")
+    
+    // has notifications for reading measurements stored in csv files on SD card
+    private let DOWNLOAD_FROM_SD_CARD_CHARACTERISTIC_UUID = CBUUID(string:"0000ffdf-0000-1000-8000-00805f9b34fb")
+    
+    private let bluetoothManager: BluetoothManager
+    
+    private var currentSessionType: SDCardSessionType?
+    private var currentSessionTypeReceived: Int = 0
+    private var currentSessionTypeExpected: Int = 0
+    
+    private var dataCharacteristicObserver: AnyHashable?
+    private var metadataCharacteristicObserver: AnyHashable?
+    
+    init(bluetoothManager: BluetoothManager) {
+        self.bluetoothManager = bluetoothManager
+    }
+    
+    func downloadData(from peripheral: CBPeripheral, progress: @escaping (SDCardData) -> Void, completion: @escaping (Result<Void, Error>) -> Void) {
+        metadataCharacteristicObserver = bluetoothManager.subscribeToCharacteristic(DOWNLOAD_META_DATA_FROM_SD_CARD_CHARACTERISTIC_UUID) { result in
+            switch result {
+            case .success(let data):
+                guard let data = data, let payload = String(data: data, encoding: .utf8) else {
+                    return
```

Comment:
Shouldn't we call `completion(.failure)` here? Otherwise, `completion` might never be called.